text/0000-futures.md

```diff
@@ -154,31 +162,315 @@ can be dropped.
 
 ### Waking up
 
-Each task executor provides its own scheduling facilities, and hence needs to
-customize the way task wakeups are handled. Most of the time, you should use the
-`std::task::Wake` trait defining wakeup behavior:
+If a future can not be directly fulfilled during execution and returns `Pending`,
+it needs a way to later on inform the executor that it needs to get polled again
+to make progress.
+
+This functionality is provided through a set of `Waker` types.
+
+`Waker`s are objects which are passed as a parameter to the `Future::poll` call,
+and which can be stored by the implementation of those `Futures`s. Whenever a
+`Future` has the need to get polled again, it can use the `wake` method of the
+waker in order to inform the executor that the task which owns the `Future`
+should get scheduled and executed again.
+
+The RFC defines two concrete `Waker` types, with which implementors of `Futures`
+and asynchronous functions will interact: `Waker` and `LocalWaker`. Both of
+these types define a `wake(&self)` function which is used to schedule the task
+that is associated to the `Waker` again.
+
+The difference between those types is that `Waker` implements the `Send` and `Sync`
+marker traits, while `LocalWaker` doesn't. This means a `Waker` can be sent to
+another thread and stored there in order to wake up the associated task later on,
+while a `LocalWaker` can't be sent. Depending on the capabilities of the underlying
+executor a `LocalWaker` can be converted into a `Waker`. Most executors in the
+ecosystem will implement this functionality. The exception will be highly
+specialized executors, which e.g. want to avoid the cost of all synchronization.
+
+Calling the `wake()` method on a `Waker` will in general be more expensive than
+on a `LocalWaker` instance, due to additional synchronization.
+
+Executors will always pass a `LocalWaker` instance to the tasks they poll.
+
+The mechanism through which tasks get scheduled again depends on the executor
+which is driving the task.
+Possible ways of waking up a an executor include:
+- If the executor is blocked on a condition variable, the condition variable
+  needs to get notified.
+- If the executor is blocked on a system call like `select`, it might need
+  to get woken up by a syscall like `write` to a pipe.
+- If the executor's thread is parked, the wakeup call needs to unpark it.
+
+In order to accomodate for this behavior, the `Waker` and `LocalWaker` types are
+defined through a `RawWaker` type, which is an struct that defines a dynamic
+dispatch mechanism, which consists of an raw object pointer and a virtual function
+pointer table (vtable). This mechanism allows implementors of an executor to
+customize the behavior of `RawWaker`, `Waker` and `LocalWaker` objects.
```

Comment:
Maybe this should be something like

> This mechanism allows implementers of an executor to customize the behavior of `Waker` and `LocalWaker` objects by providing an executor specific `RawWaker`.

Specifically noting that executors aren't directly customizing the `[Local]Waker`.

Reply:
sounds good. I will change it.